New 'inode' collector for Linux

knweiss opened the pull request:
Exports inode statistics from `/proc/sys/fs/inode-state`.

Based on the `filefd_linux.go` collector.

brian-brazil commented:
I think these metric names are likely to cause confusion with users looking for filesystem inodes rather than inodes cached in memory. Can you make them a bit more specific?

knweiss commented:
Well, I thought about `node_inode_memory_allocated` and `node_inode_memory_free` but this could be confusing, too, as these are gauges of the **number of** inodes i.e. no byte sizes.

Do you have a better suggestion?

grobie commented on collector/inode_linux.go:
```diff
@@ -0,0 +1,96 @@
+	// The inode-state proc file is separated by tabs, not spaces.
+	line := strings.Split(string(scanner.Text()), "\u0009")
+	var fileInode = map[string]string{}
+	// The inode-state proc is only 1 line with 7 values.
```
This should be validated, an invalid/unexpected file should not cause out-of-bound panics.

SuperQ commented on node_exporter.go:
```diff
@@ -31,7 +31,7 @@ import (
 
 const (
-	defaultCollectors = "conntrack,cpu,diskstats,entropy,filefd,filesystem,loadavg,mdadm,meminfo,netdev,netstat,sockstat,stat,textfile,time,uname,vmstat"
+	defaultCollectors = "conntrack,cpu,diskstats,entropy,filefd,filesystem,inode,loadavg,mdadm,meminfo,netdev,netstat,sockstat,stat,textfile,time,uname,vmstat"
```
I would suggest calling this `inode-state` to avoid confusion.

grobie commented on collector/inode_linux.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func parseInodeStats(r io.Reader, fileName string) (map[string]string, error) {
+	var scanner = bufio.NewScanner(r)
```
Using a scanner for file with a single line seems to be unnecessary. Just reading everything from the reader into a byte array should be enough. The error check of the scanner is also missing.

knweiss commented:
(Please notice that this collector is basically a clone of the existing filefd collector. I.e. all comments apply there, too.)

I've rename the collector from 'inode' to 'inodestate', added errors checks and switched from Scanner to `ioutil/ReadFile()` which simplified the code.

discordianfish commented:
Maybe node_inode_cache_allocated / node_inode_cache_free?
@brian-brazil You have some better suggestion? tbh, this is the first time I've heard about a inode cache, so not sure what the best name would be.

discordianfish commented:
Context: http://www.science.unitn.it/~fiorella/guidelinux/tlk/node110.html

So I'd rename inodestate to inode_cache.

@knweiss Sorry for the delay.. Are you still interested in getting this merged? If so, can you rename it and rebase? I'll make sure we get this merged soon.

discordianfish commented:
So this needs some work. Beside given it a proper name, it should also use ConstMetrics to avoid allocation overhead and races. Have a look at other collectors how to use it.
I'm going to close this PR for now. Feel free to open it again if you want to see this getting merged and addressed the issues.
And sorry again for the delays, we're improving the processes right now to avoid this in the future.
